pennylane/_grad.py

```diff
@@ -181,3 +182,206 @@ def _jacobian_function(*args, **kwargs):
+    """
+
+    if argnum > len(args) - 1:
+        raise ValueError(
+            "The value of 'argnum' has to be between 0 and {}; got {}".format(len(args) - 1, argnum)
+        )
+
+    x = _np.array(args[argnum])
+    gradient = _np.zeros_like(x, dtype="O")
+
+    if x.ndim == 0 and idx is not None:
+        raise ValueError(
+            "Argument {} is not an array, 'idx' should be set to 'None';"
+            " got {}".format(argnum, idx)
+        )
+
+    if idx is None:
+        idx = list(_np.ndindex(*x.shape))
+    else:
+        bounds = _np.array(x.shape) - 1
+        for _idx in idx:
+
+            if (_np.array(_idx) > bounds).any():
+                raise ValueError(
+                    "Indices in 'idx' can not be greater than {}; got {}".format(
+                        tuple(bounds), _idx
+                    )
+                )
+
+    for i in idx:
+        shift = _np.zeros_like(x)
+        shift[i] += 0.5 * delta
+        gradient[i] = (
+            f(*args[:argnum], x + shift, *args[argnum + 1 :], **kwargs)
+            - f(*args[:argnum], x - shift, *args[argnum + 1 :], **kwargs)
+        ) * delta ** -1
+
+    return gradient
+
+
+def _fd_second_order_centered(f, argnum, delta, *args, idx=None, **kwargs):
+    r"""Uses a central finite difference approximation to compute the second-order
+    derivative :math:`\frac{\partial^2 f(x)}{\partial x_i \partial x_j}` of the function ``f``
+    with respect to the argument ``argnum``.
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        argnum (int): which argument to take the derivative with respect to
+        delta (float): step size used to evaluate the finite differences
+        idx (list[int]): if argument ``args[argnum]`` is an array, `idx`` specifies
+            the indices ``i, j`` of the arguments to differentiate
+
+    Returns:
+        (float or array): the second-order derivative of the input function with respect
+        to the arguments in ``argnum``
+    """
+
+    if argnum > len(args) - 1:
+        raise ValueError(
+            "The value of 'argnum' has to be between 0 and {}; got {}".format(len(args) - 1, argnum)
+        )
+
+    x = _np.array(args[argnum])
+
+    if x.ndim == 0 and idx is not None:
+        raise ValueError(
+            "Argument {} is not an array, 'idx' should be set to 'None';"
+            " got {}".format(argnum, idx)
+        )
+
+    if idx is None:
+        idx = [(), ()]
+    else:
+        bounds = _np.array(x.shape) - 1
+        if len(idx) > 2:
+            raise ValueError(
+                "The number of indices given in 'idx' can not be greater than two; got {} indices".format(
+                    len(idx)
+                )
+            )
+
+        if (_np.array(idx) > bounds).any():
+            raise ValueError(
+                "Indices in 'idx' can not be greater than {}; got {}".format(tuple(bounds), idx)
+            )
+
+    i, j = idx
+
+    # diagonal
```

Comment:
💯 

For the hessian, it should be sufficient to just iterate over the upper triangular elements + diagonal, and then symmetrize it.

Reply:
Assuming that df/dxdy = df/dydx holds.

Reply:
Somehow the same comment here.

This function gives the second derivative of the Hamiltonian. We don't need the full Hessian of the Hamiltonian as they enter individually in the calculation of the second derivative of the energy.

Furthermore, I would not encode the calculation of the full nuclear Hessian in a function since this can be very time consuming. The same applies for the energy Hessian. In my view, If we provide the function to compute the second derivative, it is straightforward for the user to build the upper/lower triangular matrix, if the derivative is symmetric as pointed out above.

Reply:
> My original comment was a reminder, so not relevant for this PR.
> 
> Reading the [wikipedia article](https://en.wikipedia.org/wiki/Finite_difference#Multivariate_finite_differences) on finite differences I realized that when computing Hessians it's possible to be smart about not evaluating the function twice on the same points when both first and second derivatives are computed

oh, now I see. Thanks!.